crates/dbsp/src/mono/mod.rs

```diff
@@ -0,0 +1,879 @@
+//! circuit being either `RootCircuit` or `NestedCircuit`.  As such, it needs to be instantiated
+//! at most twice in any program. However, since this instantiation happens when compiling the
+//! client crate, it ends up being instantiated once in each client crate of a multi-crate project.
+//!
+//! The API in this module solves the problem by forcing the instantiation of `dyn_join` in the
+//! `dbsp` crate. It provides a less generic API: all methods take and return `OrdIndexedZSet`/`OrdZSet`
+//! streams and are specialized for a specific type of circuit, `RootCircuit` or `NestedCircuit` (which
+//! means that they don't work for arbitrary levels of nesting).  Internally they call monomorphized
+//! versions of `dyn_` operators (e.g., `dyn_join_mono`).
+//!
+//! This API is enabled by the `backend-mode` feature, which also disables the matching polymorphic
+//! API in `crate::operators`. Enabling this feature should be transparent for most client code.
+
+use std::mem::take;
+
+use crate::{
+    algebra::MulByRef,
+    circuit::WithClock,
+    dynamic::{DowncastTrait, DynData, DynUnit, DynWeight, Erase},
+    operator::{
+        dynamic::{
+            aggregate::{DynAggregatorImpl, IncAggregateFactories, IncAggregateLinearFactories},
+            controlled_filter::ControlledFilterFactories,
+            distinct::DistinctFactories,
+            join::{AntijoinFactories, JoinFactories},
+            MonoIndexedZSet,
+        },
+        join::{mk_trace_join_flatmap_funcs, mk_trace_join_funcs, mk_trace_join_generic_funcs},
+        time_series::OrdPartitionedOverStream,
+        Aggregator,
+    },
+    trace::BatchReaderFactories,
+    utils::Tup2,
+    DBData, DBWeight, DynZWeight, NestedCircuit, OrdIndexedZSet, OrdZSet, RootCircuit, Stream,
+    TypedBox, ZWeight,
+};
+
+impl<K, V> Stream<RootCircuit, OrdIndexedZSet<K, V>>
+where
+    K: DBData,
+    V: DBData,
+{
+    #[allow(clippy::type_complexity)]
+    #[track_caller]
+    pub fn aggregate<A>(&self, aggregator: A) -> Stream<RootCircuit, OrdIndexedZSet<K, A::Output>>
+    where
+        A: Aggregator<V, (), ZWeight>,
+    {
+        let aggregate_factories = IncAggregateFactories::new::<K, V, ZWeight, A::Output>();
+
+        let dyn_aggregator =
+            DynAggregatorImpl::<DynData, V, (), DynZWeight, ZWeight, A, DynData, DynData>::new(
+                aggregator,
+            );
+
+        self.inner()
+            .dyn_aggregate_mono(&aggregate_factories, &dyn_aggregator)
+            .typed()
+    }
+
+    #[track_caller]
+    pub fn aggregate_linear_postprocess<F, A, OF, OV>(
+        &self,
+        f: F,
+        of: OF,
+    ) -> Stream<RootCircuit, OrdIndexedZSet<K, OV>>
+    where
+        A: DBWeight + MulByRef<ZWeight, Output = A>,
+        OV: DBData,
+        F: Fn(&V) -> A + Clone + 'static,
+        OF: Fn(A) -> OV + Clone + 'static,
+    {
+        let factories: IncAggregateLinearFactories<
```

Comment:
this stuff is a bit numbing, hopefully the Rust type checker won't let you make any stupid mistake